pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java

```diff
@@ -179,6 +180,48 @@ public static String rtrim(String input) {
     return RTRIM.matcher(input).replaceAll("");
   }
 
+  /**
+   * @see Pattern#matches(String, CharSequence)
+   * @param value input value
+   * @param regexp regular expression
+   * @return the matched result.
+   */
+  @ScalarFunction
+  public static String regexpExtract(String value, String regexp) {
+    return regexpExtract(value, regexp, 1, "");
+  }
+
+  /**
+   * Regular expression extract that accepts starting position as argument.
+   * @param value input value
+   * @param regexp regular expression
+   * @param occurrence the specified i-th occurrence to extract
+   * @return the matched result.
+   */
+  @ScalarFunction
+  public static String regexpExtract(String value, String regexp, int occurrence) {
+    return regexpExtract(value, regexp, occurrence, "");
+  }
+
+  /**
+   * Regular expression extract that accepts starting position and i-th occurrence as argument.
+   * @param value input value
+   * @param regexp regular expression
+   * @param occurrence the specified i-th occurrence to extract
+   * @param defaultValue the default value if no match found
+   * @return the matched result
+   */
+  @ScalarFunction
+  public static String regexpExtract(String value, String regexp, int occurrence, String defaultValue) {
+    Pattern p = Pattern.compile(regexp);
+    Matcher matcher = p.matcher(value);
+    if (matcher.find()) {
+      return matcher.group(occurrence - 1);
```

Comment:
I feel this should be `occurrence` instead of `occurrence - 1`

Reply:
good catch. let search around and see what's the tradition to handle this exception.

But in SQL semantics all index starts from "1". unless Pinot follow a different one I think we should stick with 1-based indexing.

Reply:
`matcher.group(0)` will return the whole string IIRC, instead of the actual group

Reply:
I got some time to think about it. I am mixing the presto definition with the big query definition 
- presto defines the integer as the index of the capture group
- big query defines the integer as the # of times the ENTIRE regexp match occurrence. 

Another issue:
matcher.matches() also is a bit different from some behavior in other languages, for example python doesn't require the entire string to match. only the maximum first/longest match. 

I think we need to take a step back and discuss exactly what we wanted to support. I will bring this up in the original issue

Reply:
> `matcher.group(0)` will return the whole string IIRC, instead of the actual group

if I understand correctly.
```
Pattern p = Pattern.compile("([\\d]+)[\\w]+");
Matcher m = p.matcher("!!!123asdf!!!");
matcher.find(); // true
matcher.group(0); // returns "123asdf"
matcher.group(1); // returns "123"
```
where as:
```
Pattern p = Pattern.compile("([\\d]+)[\\w]+");
Matcher m = p.matcher("!!!123asdf!!!");
matcher.matches(); // false
matcher.group(0); // exception
```